src/EditorFeatures/Core/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs

```diff
@@ -217,18 +229,24 @@ private void UpdateWorkspaceForResetOfTypedIdentifier(Workspace workspace, Solut
                 localUndoTransaction.AddUndo(undoPrimitiveBefore);
 
                 if (!workspace.TryApplyChanges(newSolution))
-                {
-                    Contract.Fail("Rename Tracking could not update solution.");
-                }
+                    return (NotificationSeverity.Error, EditorFeaturesResources.Rename_operation_could_not_complete_due_to_external_change_to_workspace);
+
+                // If we successfully update the workspace then make sure the undo transaction is committed and is
+                // always able to undo anything any other external listener did.
 
                 // Never resume tracking session on redo
                 var undoPrimitiveAfter = new UndoPrimitive(_stateMachine.Buffer, trackingSessionId, shouldRestoreStateOnUndo: false);
                 localUndoTransaction.AddUndo(undoPrimitiveAfter);
 
                 localUndoTransaction.Complete();
```

Comment:
not sure waht the best logic is to have here.  previously, if applying failed, we would throw in Contract.Fail.  so this code woudl never run.  i'm preserving that logic here... which i think is desirable.  if we didn't change the workspace we don't need an undo added to the stack.

but i'm not sure.  we coudl always add these no matter what, but i'm not sure the implication of that.

we will at least add these on success, which was our logic from before.

Reply:
@jasonmalinowski ?